pkg/controllers/node/tainting/controller.go

```diff
@@ -0,0 +1,139 @@
+/*
+Copyright The Kubernetes Authors.
+
+Licensed under the Apache License, Version 2.0 (the "License");
+you may not use this file except in compliance with the License.
+You may obtain a copy of the License at
+
+    http://www.apache.org/licenses/LICENSE-2.0
+
+Unless required by applicable law or agreed to in writing, software
+distributed under the License is distributed on an "AS IS" BASIS,
+WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+See the License for the specific language governing permissions and
+limitations under the License.
+*/
+
+package tainting
+
+import (
+	"context"
+	"fmt"
+	golog "log"
+
+	"github.com/awslabs/operatorpkg/singleton"
+	"github.com/awslabs/operatorpkg/status"
+	"github.com/samber/lo"
+	v1 "k8s.io/api/core/v1"
+	"k8s.io/apimachinery/pkg/api/errors"
+	"k8s.io/apimachinery/pkg/types"
+	"k8s.io/apimachinery/pkg/util/sets"
+	controllerruntime "sigs.k8s.io/controller-runtime"
+	"sigs.k8s.io/controller-runtime/pkg/client"
+	"sigs.k8s.io/controller-runtime/pkg/controller"
+	"sigs.k8s.io/controller-runtime/pkg/log"
+	"sigs.k8s.io/controller-runtime/pkg/manager"
+	"sigs.k8s.io/controller-runtime/pkg/reconcile"
+
+	"sigs.k8s.io/karpenter/pkg/apis/v1beta1"
+
+	nodeutils "sigs.k8s.io/karpenter/pkg/utils/node"
+)
+
+var (
+	// conditionToTaintsMap maps condition values which can be present on the node to a set of taints that should be added in response
+	conditionToTaintsMap = map[string]v1.Taint{
+		v1beta1.ConditionTypeDrifted: v1beta1.DisruptionCandidatePreferNoScheduleTaint,
+		v1beta1.ConditionTypeExpired: v1beta1.DisruptionCandidatePreferNoScheduleTaint,
+		v1beta1.ConditionTypeEmpty:   v1beta1.DisruptionCandidatePreferNoScheduleTaint,
+	}
+
+	// managedTaints is the string representation of all taints managed by the tainting controller
+	managedTaints = sets.New(lo.Uniq(lo.MapToSlice(conditionToTaintsMap, func(_ string, taint v1.Taint) string {
+		return taint.ToString()
+	}))...)
+)
+
+func init() {
+	// Assert that no managed taint has a key-value collision. The tainting controller relies on this invariant, change with caution.
+	taintsPerKeyEffect := lo.GroupBy(lo.UniqBy(lo.Values(conditionToTaintsMap), func(taint v1.Taint) string {
```

Comment:
I'm removing this as part of "degeneralizing" the controller, but it actually wouldn't be possible to hit in the suite test. The idea is that this is a compile time invariant; it would fail at startup every time, or it would never fail at all. You'd have to change the data it's checking the invariant against then recompile.